### ISSUE TITLE
UI feature request: only showing top ~20 rows of pandas dataframe in variable explorer

### ISSUE THREAD

**Issue by jsh9** [spyder-ide/spyder#10003]
This is a feature request: only show top ~20 rows of pandas dataframe when opening a certain dataframe from the variable explorer, and then the user can choose to view all the rows via a click of a button.

This is to prevent Spyder from crashing or becoming very slow when opening pandas dataframes of too many rows (thousands or hundreds of thousands). 

Of course, ~20 sounds too conservative, but this is just a number off the top of my head.

**Comment by ccordoba12** [comment]
This is already in place: we only show 500 rows and 40 columns by default and more data is loaded on demand when the user reaches those limits in the viewer. However, we do need to load the entire dataframe in memory for that.

What you're proposing is to only load and display `df.head(500)` instead of the entire dataframe? That could help to improve things, but right now you can open a dataframe with 1e6 rows and 2 or 3 columns very quickly, so perhaps the problem is with really big dataframes (1e7+ rows or elements)?

For that we're thinking to use `pyarrow` to avoid the cost of making a copy of the dataframe in the kernel in order to display it in Spyder, and instead simply load it directly from memory. In other words, if you have enough memory to create a huge dataframe in your machine, with `pyarrow` you wouldn't need twice that memory to view it in Spyder.

**Comment by ccordoba12** [comment]
Closing due to lack of response.

**Comment by jsh9** [comment]
Hi @ccordoba12 , please do not close this. I have been busy for the past week.

Loading large dataframes (a few hundred columns, 100,000+ rows) is a daily routine for me and many other people. Making a copy of such large data frames are not necessary, because we usually just want to inspect what some rows and some columns look like.

**Comment by ccordoba12** [comment]
> Loading large dataframes (a few hundred columns, 100,000+ rows) is a daily routine for me and many other people

Ok, so you were talking about dataframes with 1e7+ elements, as I suspected. The problem is that implementing your suggestion would be kind of hard for us and it'd have to be removed in the future in favor of using [`pyarrow`](https://arrow.apache.org/docs/python/ipc.html#component-based-serialization).

@impact27, what do you think about this?

**Comment by impact27** [comment]
The two solutions I would see are:
 - Avoiding a copy of the data. (with pyarrow)
 - Only show part of the data to the frontend. (Why would anyone actually look at a 1e7+ pandas dataframne or numpy array in the variable explorer? Even 1e4+?) 

In my opinion, the variable explorer should be only be a way of "previewing" the data, but I am not sure what everyone use it for. In my experience:
 - scalars are fine
 - 1d arrays are fine if not too long, otherwise I am usually interested in the mean value + std + 1d plot
 - 2d array: I usually use them for images, so a plt.imshow() is usually more informative than a huge 500x500 matrix. (Even when the 2d arrays are not images, I would rather have a imshow)
 - 3d array: usually stack of 2darray
 - 4+d array: not supported anyway
 - pandas dataframe: I am usually not interested in more than 10 lines. There is some complications with multiindex, but I am usually only interested in the columns names, and the first 10 lines.

The variable explorer is useful for debugging, but I don't really see a use to edit a 1e7+ data frame in the variable explorer.

**Comment by impact27** [comment]
The easiest way to implement that would be to set a limit to the size of the variables the user can edit. Something like: "Sorry, this variable is too large to be edited through the variable explorer."

**Comment by impact27** [comment]
Or: 'Do you want to load the full variable to edit it?'

**Comment by impact27** [comment]
I guess my idea is the following: 

If the variable is small, keep the current behaviour. It is is large, only send a preview to the frontend, and disable editing.

Then, the improvment can be:
 - Give a way to the user to request the full variable to be sent over for editing. 
 - use pyarrow to avoid copies.

**Comment by ccordoba12** [comment]
> Then, the improvment can be:
>    * Give a way to the user to request the full variable to be sent over for editing.
>    * use pyarrow to avoid copies.

Yep, that's basically what @jsh9 is asking, plus my own suggestion.

The thing is we'll need to change all our viewers to handle user requests for more data, to provide a uniform interface among them.

Besides, switching to pyarrow will also require some important changes (because its only useful when frontend and kernel live in the same machine).

Therefore, we'll have to leave this for a future Spyder major version (Spyder 5 or even 6), sorry.

**Comment by impact27** [comment]
How hard would it be, for large data, to only send part of the data and disallow editing?
I think the usecase described by @jsh9 is more commun than a user editing a huge dataset?
That would allow viewing an extract of a large dataset, at the cost of not being able to edit it.

> The thing is we'll need to change all our viewers to handle user requests for more data, to provide a uniform interface among them.

The only thing we need to add in this case is a warning that the data is just a preview and can't be edited.

**Comment by impact27** [comment]
Then the other improvments can be kept for later Spyder versions.

**Comment by ccordoba12** [comment]
> Would it be easy, for large data, to only send part of the data and disallow editing?

Editing is not the problem (we already have an attribute to set the viewer to be read-only), but requesting the entire variable after having access to a preview. That's the main issue.

And with if a user asks to view a list of big dataframes mixed with other data structures? What should we do in that case? Perhaps using `pympler` in the kernel would be a good idea to not serialize that list unconditionally. But there is also the problem of trying to traverse that list to detect what elements are dataframes and from those only select a preview.

So there's a host of hard interface problems that we don't have time to consider for Spyder 4 (and I won't accept a partial solution for this at this point, i.e. when we're about to have a feature freeze for that version, sorry).

**Comment by akhtar-shah** [comment]
Hi guys, have you thought about incorporating this feature in any new releases? 
The idea that was mentioned earlier would be really good. Just display df.head(500) for large data frames.
